### PR TITLE
Create and bundle spdx and cyclonedx json SBOMs as part of stemcell creation

### PR DESCRIPTION
**Pull request by rthill91**
- Install syft in image via Dockerfile
- Add new sbom_create stage to generate sbom.spdx.json as part of stemcell creation
- Bundle sbom.spdx.json into resulting stemcell tgz
- Bundle sbom.cdx.json into resulting stemcell tgz

### CODE REVIEW

File: ci/docker/os-image-stemcell-builder-jammy/Dockerfile [hunk 1]
```diff
@@ -56,6 +56,9 @@ RUN groupadd -o -g ${GROUP_ID} ubuntu                  \
   && useradd -u ${USER_ID} -g ${GROUP_ID} -m ubuntu \
   && echo 'ubuntu ALL=NOPASSWD:ALL' >> /etc/sudoers
 
+# install syft for sbom generation
+RUN curl -sSfL https://raw.githubusercontent.com/anchore/syft/main/install.sh | sh -s -- -b /usr/local/bin
```
Comment (rkoster) [inline ci/docker/os-image-stemcell-builder-jammy/Dockerfile]: Please use an immutable URL and digest verification.

**Review by rkoster — CHANGES_REQUESTED** [review]
PR looks great! Not sure if we want to do some filtering to reduce false positives (at some point the google agent tools had a dev dependency which would be picked up by syft). But those could also be done at a later point in time, once people start consuming the sbom and provide feedback.

**Comment by ragaskar** [comment]
I agree w/ @rkoster that there should be some provision for checking the provenance of the install script/installed libraries. 

I can also share that in our experience, we've seen issues with tracking latest `syft` in that new features are added very quickly and occasionally these new features have undesirable impacts (e.g., we've seen new cataloguers suddenly introduce records for dozens of libraries that -- in our case -- were all false positives). It might be nice to have a way to easily pin syft when necessary. It's possible there is some implementation overlap with provenance for version pinning, so perhaps it makes sense to take that on now.

**Review by ragaskar — DISMISSED** [review]
See my comment above, I see @rkoster is already asking for some validation strategy with the syft install, which I agree with. 

I don't think the version-pinning provision I mentioned is required so I'll give this a thumbs up, but it's a definite nice-to-have that may intersect with the provenance piece.

**Comment by ragaskar** [comment]
also, thanks for this contribution, cool stuff!

File: ci/docker/os-image-stemcell-builder-jammy/Dockerfile [hunk 2]
```diff
@@ -56,6 +56,15 @@ RUN groupadd -o -g ${GROUP_ID} ubuntu                  \
...
   && echo 'ubuntu ALL=NOPASSWD:ALL' >> /etc/sudoers
 
+# install syft for sbom generation
+ADD syft.checksum /tmp/syft.checksum
+ADD https://raw.githubusercontent.com/anchore/syft/5a865d0d904b90252c02e0cceeac61c3d882837d/install.sh /tmp/install.sh
+RUN \
+    cd tmp \
+    sha256sum -c --status syft.checksum && \
+    sh install.sh -b /usr/local/bin && \
+    rm install.sh syft.checksum
```
Comment (rkoster) [inline ci/docker/os-image-stemcell-builder-jammy/Dockerfile]: I would prefer a simple one-liner like: https://github.com/cloudfoundry/bosh-linux-stemcell-builder/blob/4481987e71b816ae0bf7f20119adc3aa7f610cf4/ci/docker/os-image-stemcell-builder-jammy/Dockerfile#L113-L116 over downloading and executing an external install script.

**Comment by beyhan** [comment]
What is the estimated size of the generated SBOM? Have you also tried the other formats supported like the cycloneDX and the syft own format?

**Comment by rthill91** [comment]
> What is the estimated size of the generated SBOM? Have you also tried the other formats supported like the cycloneDX and the syft own format?

~34M in my jammy fips test build from today.

As for other formats, minimal consideration was given beyond the desire for something generic and machine readable. That put syft format out. Beyond that, internally spdx is the chosen format.

**Review by rkoster — DISMISSED** [review]

**Comment by VinodAnandan** [comment]
@rthill91 Could you please consider adding support for CycloneDX format as well?

**Comment by rkoster** [comment]
Our internal tooling uses CycloneDX, so supporting both formats would be a plus for us.
The `cyclonedx-json` output format [seems to be supported by syft](https://github.com/anchore/syft?tab=readme-ov-file#output-formats).
But this is not a blocker for me and could be done by someone else in a later PR. I just want to make sure the discussion about supporting multiple formats is captured here.

**Comment by rthill91** [comment]
> Our internal tooling uses CycloneDX, so supporting both formats would be a plus for us. The `cyclonedx-json` output format [seems to be supported by syft](https://github.com/anchore/syft?tab=readme-ov-file#output-formats). But this is not a blocker for me and could be done by someone else in a later PR. I just want to make sure the discussion about supporting multiple formats is captured here.

syft definitely supports cyclonedx. The only complexity (off the top of my head) to adding support is that it makes it turns sbom generation into a configurable option rather than a static thing and it'll make the expected files in the output tgz configurable. I could just change the sbom to `sbom.json` to avoid that last bit, but I like the clarity of specifying that it's spdx, particularly if there's support for other formats included.

I don't mind working this, but it'll likely have to wait a few weeks (Noble release is going to keep me busy) and I'd rather it be a separate PR.

**Comment by rkoster** [comment]
I would not make it an optional build flag, but instead just include both sboms, similar to what buildpacks are doing: https://github.com/buildpacks/rfcs/blob/main/text/0095-sbom.md#what-it-is

So we would end up with something like:
```ruby
expected = ['stemcell.MF', 'packages.txt', 'dev_tools_file_list.txt', 'image', 'sbom.spdx.json', 'sbom.cdx.json']
```

**Comment by rthill91** [comment]
> I would not make it an optional build flag, but instead just include both sboms, similar to what buildpacks are doing: https://github.com/buildpacks/rfcs/blob/main/text/0095-sbom.md#what-it-is
> 
> So we would end up with something like:
> 
> ```ruby
> expected = ['stemcell.MF', 'packages.txt', 'dev_tools_file_list.txt', 'image', 'sbom.spdx.json', 'sbom.cdx.json']
> ```

including both makes this a trivial change. I have it staged locally but I'm seeing some test failures (going through the non-iaas-kernel tests rather than aws specific kernel). I'll have to look at this a bit later

**Comment by rthill91** [comment]
updated to generate cyclonedx and spdx sboms

File: bosh-stemcell/lib/bosh/stemcell/stemcell_packager.rb [hunk 3]
```diff
@@ -71,7 +71,7 @@ def create_tarball(disk_format)
         stemcell_name = ArchiveFilename.new(version, definition, 'bosh-stemcell', disk_format).to_s
         tarball_name = File.join(tarball_path, stemcell_name)
 
-        expected = ['stemcell.MF', 'packages.txt', 'dev_tools_file_list.txt', 'image']
+        expected = ['stemcell.MF', 'packages.txt', 'dev_tools_file_list.txt', 'image', 'sbom.spdx.json', 'sbom.cyclonedx.json']
```
Comment (rkoster) [inline bosh-stemcell/lib/bosh/stemcell/stemcell_packager.rb]: According to: https://cyclonedx.org/specification/overview/#recognized-file-patterns
the filename should match `*.cdx.json`

**Review by rkoster — CHANGES_REQUESTED** [review]

**Review by rkoster — APPROVED** [review]

**Comment by rthill91** [comment]
updated title/description to reflect the addition of cdx sbom

**Review by ragaskar — APPROVED** [review]
thank you, looks great.

**Comment by rkoster** [comment]
Thanks! @rthill91

**Pull request merged**